### DISCUSSION TITLE
how to get extreme ray for unbounded lp?

### DISCUSSION THREAD

**Discussion by Miraclelucy** [google/or-tools#4248] [Linear Solver questions]
so Has it been resolved?
*  https://github.com/google/or-tools/issues/653

**Comment by soppera** [comment]
Hi,

You can get rays with the MathOpt API; for example with Glop:

```py
model = mathopt.Model(name='unbounded LP')
x = model.add_variable(name='x')
y = model.add_variable(name='y')
model.add_linear_constraint(x + y == 5)
model.maximize(x)

result = mathopt.solve(model, mathopt.SolverType.GLOP,
                       params=mathopt.SolveParameters(
                           presolve=mathopt.Emphasis.OFF,
                           scaling=mathopt.Emphasis.OFF,
                           lp_algorithm=mathopt.LPAlgorithm.PRIMAL_SIMPLEX))
assert result.termination.reason == mathopt.TerminationReason.UNBOUNDED
print(result.primal_rays)
```

will print

```
[PrimalRay(variable_values={<Variable id: 0, name: 'x'>: 1.0, <Variable id: 1, name: 'y'>: -1.0})]
```

Note though that getting rays requires setting properly the solver settings so that it produces them. For example with presolver and scaling enabled (which are by default) on Glop you don't get rays. Usually our unit tests show the proper settings to get rays, for example [here](https://github.com/google/or-tools/blob/870edf6f7bff6b8ff0d267d936be7e331c5b8c2d/ortools/math_opt/solvers/glop_solver_test.cc#L50-L61) for the test of Glop.
